api/src/main/java/org/apache/iceberg/io/RetryableInputStream.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.io;
+
+import dev.failsafe.Failsafe;
+import dev.failsafe.RetryPolicy;
+import dev.failsafe.RetryPolicyBuilder;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.SocketTimeoutException;
+import java.time.Duration;
+import java.util.List;
+import java.util.function.Supplier;
+import javax.net.ssl.SSLException;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+/**
+ * RetryableInputStream wraps over an underlying InputStream and retries failures encountered when
+ * reading through the stream. On retries, the underlying streams will be reinitialized.
+ */
+public class RetryableInputStream extends InputStream {
+
+  private final InputStream underlyingStream;
+  private final RetryPolicy<Object> retryPolicy;
+
+  private RetryableInputStream(InputStream underlyingStream, RetryPolicy<Object> retryPolicy) {
+    this.underlyingStream = underlyingStream;
+    this.retryPolicy = retryPolicy;
+  }
+
+  @Override
+  public int read() throws IOException {
+    return Failsafe.with(retryPolicy).get(() -> underlyingStream.read());
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    return Failsafe.with(retryPolicy).get(() -> underlyingStream.read(b, off, len));
+  }
+
+  @Override
+  public void close() throws IOException {
+    underlyingStream.close();
+  }
+
+  public static RetryableInputStream.Builder builderFor(Supplier<InputStream> newStreamSupplier) {
+    return new Builder(newStreamSupplier);
+  }
+
+  public static class Builder {
+
+    private InputStream underlyingStream;
+    private final Supplier<InputStream> newStreamProvider;
+    private List<Class<? extends Exception>> retryableExceptions =
+        ImmutableList.of(SSLException.class, SocketTimeoutException.class);
```

Comment:
[QQ] 
- Should we retry this 5xx errors like 503 ?
- will it make sense to make this configurable by any chance ? 
- Also any thoughts about outputstream ?

Reply:
@singhpk234 I believe the SDK already handles 503 errors and other errors that happen during the initiation of the request.  I believe output stream is different (I don't think we've heard of issues on that side), but I would handle that in a separate PR if necessary.

Reply:
> I believe the SDK already handles 503 errors and other errors that happen during the initiation of the request

yup, they do, but we have defaults of SDK at the moment right ? wondering if we should either expose conf's to overwrite it or control it via here. 

Also was wondering if we can go as exhaustive a retry policy and have it at one place, as this : https://github.com/apache/hadoop/blob/b108e9e2d814cc009c4b663799e056a65bf24766/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3ARetryPolicy.java#L193

In past we have been reached for tickets like this : https://github.com/apache/iceberg/issues/9124#issuecomment-1821822583


>  I believe output stream is different (I don't think we've heard of issues on that side), but I would handle that in a separate PR if necessary

+1